symengine/parser/parser.cpp

```diff
@@ -283,20 +283,15 @@ Parser::parse_implicit_mul(const std::string &expr)
     char *endptr = 0;
     std::strtod(startptr, &endptr);
 
-    RCP<const Basic> num = one, sym;
-
     // Numerical part of the result of e.g. "100x";
     size_t length = endptr - startptr;
     std::string lexpr = std::string(startptr, length);
-    num = parse_numeric(lexpr);
+    RCP<const Basic> num = parse_numeric(lexpr);
 
     // get the rest of the string
     lexpr = std::string(startptr + length, expr.length() - length);
-    if (lexpr.length() == 0) {
-        sym = one;
-    } else {
-        sym = parse_identifier(lexpr);
-    }
+    SYMENGINE_ASSERT(lexpr.length() > 0);
```

Comment:
I'd rather this throws an exception even in Release mode instead of silently ignoring.

Reply:
Can you give an example when this can happen? I couldn't figure out any.

Reply:
And if we cannot figure out when this could happen, then the assert statement is appropriate. Asserts are for ensuring that our assumptions in the code are consistent/valid, but those do not run in release mode, because we assume our code is correct in release mode.

Reply:
If I remember correctly, there was some ambiguity in the grammar where 2e10 was parsed as an implicit mul.

Reply:
I'll see if I can trigger it. Too bad there wasn't a test for this. All parsing tests pass. We need to test this.

Reply:
Quite frankly, I don't like this implicit mul parsing, due to such ambiguities. Do you think people want implicit mul to be parsed?

Reply:
> Do you think people want implicit mul to be parsed?

Yes, julia people use this feature heavily in SymEngine.jl

Reply:
FWIW, I'm no fan of implicit mul.

Reply:
While we are at it, we should also fix https://github.com/symengine/symengine/issues/1462. We can even have a few different parsers if needed.